File: source/cpp/lib/oa/ml/rollout.cpp
#include <oa/ml/rollout.h>

#include <oa/core/bufferAccess.h>
#include <oa/core/fnMatrix.h>
#include <oa/runtime/executionSession.h>
#include <oa/core/std/algo.h>
#include <oa/core/std/limits.h>

namespace {

oa::MatrixShape prefixShape(
	oa::U32 inFirst,
	const oa::MatrixShape& inSuffix) {
	oa::MatrixShape shape;
	shape.rank = inSuffix.rank + 1;
	shape[0] = static_cast<oa::I64>(inFirst);
	for (oa::I32 index = 0; index < inSuffix.rank; ++index) {
		shape[index + 1] = inSuffix[index];
	}
	return shape;
}

oa::MatrixShape prefixShape(
	oa::U32 inFirst,
	oa::U32 inSecond,
	const oa::MatrixShape& inSuffix) {
	oa::MatrixShape shape;
	shape.rank = inSuffix.rank + 2;
	shape[0] = static_cast<oa::I64>(inFirst);
	shape[1] = static_cast<oa::I64>(inSecond);
	for (oa::I32 index = 0; index < inSuffix.rank; ++index) {
		shape[index + 2] = inSuffix[index];
	}
	return shape;
}

bool isF32Vector(const oa::Matrix& inMatrix, oa::U32 inEnvironments) {
	return inMatrix.getDtype() == oa::ScalarType::Float32
		&& inMatrix.getShape()
			== oa::MatrixShape{static_cast<oa::I64>(inEnvironments)};
}

bool isU8Vector(const oa::Matrix& inMatrix, oa::U32 inEnvironments) {
	return inMatrix.getDtype() == oa::ScalarType::UInt8
		&& inMatrix.getShape()
			== oa::MatrixShape{static_cast<oa::I64>(inEnvironments)};
}

} // namespace

bool oa::RolloutBatch::isValid() const noexcept {
	return !observation.isEmpty() && !action.isEmpty() && !reward.isEmpty()
		&& !value.isEmpty() && !nextValue.isEmpty()
		&& !oldLogProbability.isEmpty() && !terminated.isEmpty()
		&& !truncated.isEmpty() && !valid.isEmpty()
		&& !advantage.isEmpty() && !ret.isEmpty();
}

oa::Result<oa::RolloutBuffer> oa::RolloutBuffer::create(
	const oa::RolloutConfig& inConfig) {
	if (inConfig.time == 0 || inConfig.environments == 0
		|| inConfig.observationShape.rank < 1
		|| inConfig.observationShape.rank > OA_MAX_TENSOR_DIMS - 2) {
		return oa::Status::invalidArgument(
			"oa::RolloutBuffer::create expects non-zero time/environments and an observation rank in [1,6]");
	}
	for (oa::I32 dim = 0; dim < inConfig.observationShape.rank; ++dim) {
		if (inConfig.observationShape[dim] <= 0) {
			return oa::Status::invalidArgument(
				"oa::RolloutBuffer::create observation dimensions must be positive");
		}
	}
	oa::U64 observationElements = 1;
	for (oa::I32 dim = 0; dim < inConfig.observationShape.rank; ++dim) {
		const oa::U64 dimension = static_cast<oa::U64>(inConfig.observationShape[dim]);
		if (observationElements
			> static_cast<oa::U64>(oa::Limits<oa::U32>::max()) / dimension) {
			return oa::Status::error(
				oa::StatusCode::OutOfRange,
				"oa::RolloutBuffer::create observation size exceeds the current 32-bit GPU indexing limit");
		}
		observationElements *= dimension;
	}
	const oa::U64 rolloutElements = static_cast<oa::U64>(inConfig.time)
		* inConfig.environments;
	if (rolloutElements > oa::Limits<oa::U32>::max()
		|| observationElements
			> static_cast<oa::U64>(oa::Limits<oa::U32>::max())
				/ rolloutElements) {
		return oa::Status::error(
			oa::StatusCode::OutOfRange,
			"oa::RolloutBuffer::create exceeds the current 32-bit GPU indexing limit");
	}

	oa::RolloutBuffer result;
	result.config_ = inConfig;
	result.observationElements_ = static_cast<oa::U32>(observationElements);
	const oa::MatrixShape scalarShape{
		static_cast<oa::I64>(inConfig.time),
		static_cast<oa::I64>(inConfig.environments)};
	result.batch_ = oa::RolloutBatch{
		.observation = oa::FnMatrix::empty(prefixShape(
			inConfig.time, inConfig.environments, inConfig.observationShape),
			oa::ScalarType::Float32),
		.action = oa::FnMatrix::empty(scalarShape, oa::ScalarType::Int32),
		.reward = oa::FnMatrix::empty(scalarShape, oa::ScalarType::Float32),
		.value = oa::FnMatrix::empty(scalarShape, oa::ScalarType::Float32),
		.nextValue = oa::FnMatrix::empty(scalarShape, oa::ScalarType::Float32),
		.oldLogProbability = oa::FnMatrix::empty(
			scalarShape, oa::ScalarType::Float32),
		.terminated = oa::FnMatrix::empty(scalarShape, oa::ScalarType::UInt8),
		.truncated = oa::FnMatrix::empty(scalarShape, oa::ScalarType::UInt8),
		.valid = oa::FnMatrix::empty(scalarShape, oa::ScalarType::UInt8),
		.advantage = oa::FnMatrix::empty(scalarShape, oa::ScalarType::Float32),
		.ret = oa::FnMatrix::empty(scalarShape, oa::ScalarType::Float32),
	};
	if (!result.batch_.isValid()) {
		return oa::Status::error(
			oa::StatusCode::OutOfMemory,
			"oa::RolloutBuffer::create could not allocate rollout storage");
	}
	return result;
}

oa::Status oa::RolloutBuffer::append(const oa::RolloutTransition& inTransition) {
	if (!isValid()) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"oa::RolloutBuffer::append requires a valid buffer");
	}
	if (finalized_) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"oa::RolloutBuffer::append requires reset after finalize");
	}
	if (isFull()) {
		return oa::Status::error(
			oa::StatusCode::ResourceExhausted,
			"oa::RolloutBuffer::append exceeds rollout capacity");
	}
	const oa::MatrixShape observationShape = prefixShape(
		config_.environments, config_.observationShape);
	const oa::MatrixShape actionShape{static_cast<oa::I64>(config_.environments)};
	const bool valid = inTransition.observation.getShape() == observationShape
		&& inTransition.observation.getDtype() == oa::ScalarType::Float32
		&& inTransition.action.getShape() == actionShape
		&& inTransition.action.getDtype() == oa::ScalarType::Int32
		&& isF32Vector(inTransition.reward, config_.environments)
		&& isF32Vector(inTransition.value, config_.environments)
		&& isF32Vector(inTransition.nextValue, config_.environments)
		&& isF32Vector(inTransition.logProbability, config_.environments)
		&& isU8Vector(inTransition.terminated, config_.environments)
		&& isU8Vector(inTransition.truncated, config_.environments);
	if (!valid) {
		return oa::Status::error(
			oa::StatusCode::ShapeMismatch,
			"oa::RolloutBuffer::append transition does not match the configured categorical rollout shape/dtype contract");
	}

	struct Push {
		oa::U32 step;
		oa::U32 environments;
		oa::U32 observationElements;
	} push{
		.step = size_,
		.environments = config_.environments,
		.observationElements = observationElements_,
	};
	oa::BufferAccess access[] = {
		oa::BufferAccess::Read, oa::BufferAccess::Read,
		oa::BufferAccess::Read, oa::BufferAccess::Read,
		oa::BufferAccess::Read, oa::BufferAccess::Read,
		oa::BufferAccess::Read, oa::BufferAccess::Read,
		oa::BufferAccess::Write, oa::BufferAccess::Write,
		oa::BufferAccess::Write, oa::BufferAccess::Write,
		oa::BufferAccess::Write, oa::BufferAccess::Write,
		oa::BufferAccess::Write, oa::BufferAccess::Write,
		oa::BufferAccess::Write,
	};
	const oa::U32 workItems = oa::max(
		config_.environments * observationElements_, config_.environments);
	oa::ExecutionSession::getActive().add(
		"RlRolloutAppend",
		{&inTransition.observation, &inTransition.action,
		 &inTransition.reward, &inTransition.value,
		 &inTransition.nextValue, &inTransition.logProbability,
		 &inTransition.terminated, &inTransition.truncated,
		 &batch_.observation, &batch_.action,
		 &batch_.reward, &batch_.value, &batch_.nextValue,
		 &batch_.oldLogProbability, &batch_.terminated,
		 &batch_.truncated, &batch_.valid},
		access,
		&push,
		sizeof(push),
		(workItems + 255U) / 256U);
	++size_;
	return oa::Status::ok();
}

oa::Status oa::RolloutBuffer::finalize(const oa::GaeConfig& inConfig) {
	if (!isValid()) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"oa::RolloutBuffer::finalize requires a valid buffer");
	}
	if (finalized_) return oa::Status::ok();
	if (!isFull()) {
		return oa::Status::error(
			oa::StatusCode::FailedPrecondition,
			"oa::RolloutBuffer::finalize requires a complete rollout");
	}
	const oa::Status status = oa::FnAdvantage::gaeInto(
		batch_.reward, batch_.value, batch_.nextValue,
		batch_.terminated, batch_.truncated,
		batch_.advantage, batch_.ret, inConfig);
	if (status.isError()) return status;
	finalized_ = true;
	return oa::Status::ok();
}

void oa::RolloutBuffer::reset() {
	if (!isValid()) return;
	struct Push { oa::U32 count; } push{
		.count = static_cast<oa::U32>(batch_.valid.numElements())};
	oa::BufferAccess access[] = {oa::BufferAccess::Write};
	oa::ExecutionSession::getActive().add(
		"RlRolloutReset", {&batch_.valid}, access,
		&push, sizeof(push), (push.count + 255U) / 256U);
	size_ = 0;
	finalized_ = false;
}

void oa::RolloutBuffer::abortUnsubmitted() noexcept {
	size_ = 0;
	finalized_ = false;
}

File: src/rs/ml/lowering/rollout.rs
//! Private lowering for fixed-capacity rollout storage mutations.

use crate::ml::rollout::{RolloutBatch, RolloutTransition};
use crate::runtime::{BufferBinding, KernelId, PushConstant};
use crate::{DType, Error, Matrix, OpAttribute, Result};

use super::common::{record_semantic, shader_u32};

pub(in crate::ml) fn append(
	transition: &RolloutTransition,
	batch: &RolloutBatch,
	step: usize,
	environments: usize,
	observation_shape: &[usize],
	observation_elements: u32,
) -> Result<()> {
	const OPERATION: &str = "oa::ml::rollout::append";
	let mut expected_observation_shape = Vec::with_capacity(observation_shape.len() + 1);
	expected_observation_shape.push(environments);
	expected_observation_shape.extend_from_slice(observation_shape);
	let vector_shape = [environments];
	let fields = [
		(
			transition.observation(),
			DType::F32,
			expected_observation_shape.as_slice(),
		),
		(transition.action(), DType::I32, vector_shape.as_slice()),
		(transition.reward(), DType::F32, vector_shape.as_slice()),
		(transition.value(), DType::F32, vector_shape.as_slice()),
		(transition.next_value(), DType::F32, vector_shape.as_slice()),
		(
			transition.log_probability(),
			DType::F32,
			vector_shape.as_slice(),
		),
		(transition.terminated(), DType::U8, vector_shape.as_slice()),
		(transition.truncated(), DType::U8, vector_shape.as_slice()),
	];
	for (matrix, dtype, shape) in &fields {
		if matrix.dtype() != *dtype || matrix.shape() != *shape {
			return Err(Error::invalid_argument(format!(
				"{OPERATION} transition does not match the configured categorical shape/dtype contract"
			)));
		}
		if !transition
			.observation()
			.engine_handle()
			.same_as(matrix.engine_handle())
		{
			return Err(Error::invalid_argument(format!(
				"{OPERATION} transition fields must belong to one engine"
			)));
		}
	}
	let destinations = batch_destinations(batch);
	for destination in destinations {
		if !transition
			.observation()
			.engine_handle()
			.same_as(destination.engine_handle())
		{
			return Err(Error::invalid_argument(format!(
				"{OPERATION} transition and rollout buffer must belong to one engine"
			)));
		}
		if fields
			.iter()
			.any(|(source, _, _)| source.storage().same_as(destination.storage()))
		{
			return Err(Error::invalid_argument(format!(
				"{OPERATION} transition fields must not alias retained rollout storage"
			)));
		}
	}

	let step = shader_u32(step, "step", OPERATION)?;
	let environments = shader_u32(environments, "environment count", OPERATION)?;
	let observation_count = environments
		.checked_mul(observation_elements)
		.ok_or_else(|| {
			Error::resource_exhausted(format!("{OPERATION} observation count exceeds u32"))
		})?;
	let work_items = observation_count.max(environments);
	let inputs = [
		transition.observation(),
		transition.action(),
		transition.reward(),
		transition.value(),
		transition.next_value(),
		transition.log_probability(),
		transition.terminated(),
		transition.truncated(),
		batch.observation(),
		batch.action(),
		batch.reward(),
		batch.value(),
		batch.next_value(),
		batch.old_log_probability(),
		batch.terminated(),
		batch.truncated(),
		batch.valid(),
	];
	let outputs = batch_destinations(batch);
	let buffers = [
		BufferBinding::read(transition.observation().storage()),
		BufferBinding::read(transition.action().storage()),
		BufferBinding::read(transition.reward().storage()),
		BufferBinding::read(transition.value().storage()),
		BufferBinding::read(transition.next_value().storage()),
		BufferBinding::read(transition.log_probability().storage()),
		BufferBinding::read(transition.terminated().storage()),
		BufferBinding::read(transition.truncated().storage()),
		BufferBinding::write(batch.observation().storage()),
		BufferBinding::write(batch.action().storage()),
		BufferBinding::write(batch.reward().storage()),
		BufferBinding::write(batch.value().storage()),
		BufferBinding::write(batch.next_value().storage()),
		BufferBinding::write(batch.old_log_probability().storage()),
		BufferBinding::write(batch.terminated().storage()),
		BufferBinding::write(batch.truncated().storage()),
		BufferBinding::write(batch.valid().storage()),
	];
	let push_constants = [
		PushConstant::U32(step),
		PushConstant::U32(environments),
		PushConstant::U32(observation_elements),
	];
	let attributes = [
		OpAttribute::UnsignedInteger {
			name: "step".into(),
			value: u64::from(step),
		},
		OpAttribute::UnsignedInteger {
			name: "environments".into(),
			value: u64::from(environments),
		},
		OpAttribute::UnsignedInteger {
			name: "observation_elements".into(),
			value: u64::from(observation_elements),
		},
	];
	let kernel = KernelId::MlRolloutAppendF32;
	record_semantic(
		&inputs,
		&outputs,
		&attributes,
		kernel,
		&buffers,
		&push_constants,
		kernel.linear_workgroups(work_items),
	)
}

pub(in crate::ml) fn reset(valid: &Matrix) -> Result<()> {
	const OPERATION: &str = "oa::ml::rollout::reset";
	if valid.dtype() != DType::U8 || valid.shape().len() != 2 {
		return Err(Error::invalid_argument(format!(
			"{OPERATION} requires a rank-two U8 validity mask"
		)));
	}
	let count = shader_u32(valid.num_elements(), "element count", OPERATION)?;
	let buffers = [BufferBinding::write(valid.storage())];
	let push_constants = [PushConstant::U32(count)];
	let kernel = KernelId::MlRolloutResetU8;
	record_semantic(
		&[valid],
		&[valid],
		&[],
		kernel,
		&buffers,
		&push_constants,
		kernel.linear_workgroups(count),
	)
}

fn batch_destinations(batch: &RolloutBatch) -> [&Matrix; 9] {
	[
		batch.observation(),
		batch.action(),
		batch.reward(),
		batch.value(),
		batch.next_value(),
		batch.old_log_probability(),
		batch.terminated(),
		batch.truncated(),
		batch.valid(),
	]
}
